Fix dropdown escape propagation

alpadev opened the pull request:
Closes #33465

alpadev committed:
refactor: extract the menu item selection logic into its own function

this reduces the complexity of the dataApiKeydownHandler to get rid of the eslint warning

alpadev committed:
test: add test case for dropdown escape key propagation

alpadev committed:
fix: escape key should not be handled when the dropdown is closed

see #33465

GeoSot commented:
LGTM

rohit2sharma95 commented on js/src/dropdown.js:
```diff
@@ -463,26 +488,29 @@ class Dropdown extends BaseComponent {
 
-    const parent = Dropdown.getParentFromElement(this)
-    const isActive = this.classList.contains(CLASS_NAME_SHOW)
+    const getButton = () => this.matches(SELECTOR_DATA_TOGGLE) ? this : SelectorEngine.prev(this, SELECTOR_DATA_TOGGLE)[0]
```
This may be better :slightly_smiling_face: 
`getButton` => `getToggleButton`

rohit2sharma95 reviewed (commented):
https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L489-L492

This should not be changed as below? :thinking: 
```js
if (isActive && event.key === SPACE_KEY) {
  Dropdown.clearMenus()
  return
}
```

alpadev committed:
refactor: changed getButton to getToggleButton

requested by rohit2sharma95 - https://github.com/twbs/bootstrap/pull/33479#discussion_r603857306

alpadev commented:
> https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L489-L492
> 
> This should not be changed as below? 🤔
> 
> ```js
> if (isActive && event.key === SPACE_KEY) {
>   Dropdown.clearMenus()
>   return
> }
> ```

Not sure if this is even needed.

This checks for keys that are dropdown commands.

https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L458-L464

But the regex is missing the space key so I'm not sure if the code would even reach this block.

https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L44

rohit2sharma95 commented:
This change was made in #29885. Anything related to input groups?

/CC @MartijnCuppens

rohit2sharma95 commented:
And can not the `selectMenuItem` method be a private method instead of a static one (as it is not being used directly :thinking:)? It can be used through the context of the dropdown then IMO. Because the `dataApiKeydownHandler` method is bound with the dropdown elements only (menu and the toggle button).

alpadev commented:
> And can not the `selectMenuItem` method be a private method instead of a static one (as it is not being used directly 🤔)? It can be used through the context of the dropdown then IMO. Because the `dataApiKeydownHandler` method is bound with the dropdown elements only (menu and the toggle button).

I didn't mean to refactor too much in the scope of this issue but rather to just get rid of that Eslint warning. I agree that using a static method isn't ideal but the item slection functionality wasn't bound to the instance before and I made it the way `getParentFromElement` was implemented.

If you want me to change it I should be able to do this tho.

GeoSot commented:
Till now is just a refactoring that I can follow, a minor fix and a valid test that succeeds . So **till here** I find it **OK**.

I agree to refactor `selectMenuItem` to private method, and would also love to see `dataApiKeydownHandler` gordian knot untied, but on another PR, to avoid mistakes

So if we can all agree on these, I would prefer to keep it as it is and if @alpadev has the courage to keep up with a following PR

XhmikosR committed:
Merge branch 'main' into fix-dropdown-escape-propagation
